### DISCUSSION TITLE
How to handle Errors from server actions into client side component

### DISCUSSION THREAD

**Discussion by suhejbfejziu** [vercel/next.js#73399] [App Router]
I'm currently using external APIs to perform mutations with server actions. However, when a mutation occurs, I need to display a message or show a toast notification in the client component. The issue arises when I throw an error within the server actions—it consistently results in a 500 Internal Server Error. This doesn't seem like the best approach for handling errors. Should I consider switching to something like TanStack Query for mutations, or is there a better way to handle errors and communicate them effectively to the client?

**Comment by denis23x** [comment]
Hi @suhejbfejziu 

You can set cookies on server side, check them on client, and display message.

**Comment by icyJoseph** [comment]
The problem is that server actions, as I see it, expect you to handle your business scenarios, so that your server action always succeeds, and can communicate back to the client, what the result was. You are literally triggering server side code through calling a server action, and a failure on that, is too open ended. 

Instead of throwing an error, you could return a object, with a given shape that, your client can interpret as necessary.

- https://react.dev/reference/rsc/use-server#serializable-parameters-and-return-values

**Reply by suhejbfejziu** [reply]
So you are saying that I should never throw any errors in server actions just to work as bridge to get response just return directly response without try catch?
